### PR TITLE
[TS migration] Migrate str to Typescript (expensify-common)

### PR DESCRIPTION
**Pull request by blazejkustra**
The PR aims to migrate `Str` module to Typescript, it also removes almost all underscore usage from str and instead uses native JS.

### Fixed Issues
$ [GH_LINK](https://github.com/Expensify/App/issues/42246)

### CODE REVIEW

File: lib/str.ts [hunk 1]
```diff
@@ -212,11 +207,11 @@ const Str = {
      * The ellipses will only be appended if @p text is longer than the @p length
      * given.
      *
-     * @param {String} val   The string to reduce in size.
-     * @param {Number} length The maximal length desired.
-     * @return {String} The shortened @p text.
+     * @param val   The string to reduce in size.
```
Comment (Skalakid) [inline lib/str.ts:210]:
```suggestion
     * @param val The string to reduce in size.
```

File: lib/str.ts [hunk 2]
```diff
@@ -109,31 +112,31 @@ const Str = {
     /**
      * Escape text while preventing any sort of double escape, so 'X & Y' -> 'X &amp; Y' and 'X &amp; Y' -> 'X &amp; Y'
      *
-     * @param {String} s the string to escape
-     * @return {String} the escaped string
+     * @param s the string to escape
+     * @return string the escaped string
```
Comment (Skalakid) [inline lib/str.ts:116]: Let's be consistent and start descriptions after variable names with capital letter in all functions
```suggestion
     * @param s The string to escape
     * @returns the escaped string
```

File: lib/str.ts [hunk 3]
```diff
@@ -109,31 +112,31 @@ const Str = {
...
 
     /**
      * HTML encoding insensitive equals.
      *
-     * @param {String} first string to compare
-     * @param {String} second string to compare
-     * @return {Boolean} true when first === second, ignoring HTML encoding
+     * @param first string to compare
+     * @param second string to compare
+     * @return boolean true when first === second, ignoring HTML encoding
```
Comment (Skalakid) [inline lib/str.ts:127]:
```suggestion
     * @param first string to compare
     * @param second string to compare
     * @returns boolean true when first === second, ignoring HTML encoding
```

File: lib/str.ts [hunk 4]
```diff
@@ -25,37 +40,28 @@ const Str = {
...
      * Converts a USD string into th number of cents it represents.
      *
-     * @param {String}  amountStr     A string representing a USD value.
-     * @param {Boolean} allowFraction Flag indicating if fractions of cents should be
+     * @param amountStr A string representing a USD value.
+     * @param allowFraction Flag indicating if fractions of cents should be
      *                               allowed in the output.
      *
-     * @return {Number} The cent value of the @p amountStr.
+     * @return number The cent value of the @p amountStr.
```
Comment (Skalakid) [inline lib/str.ts:47]:
```suggestion
     * @returns The cent value of the @p amountStr.
```

File: lib/str.ts [hunk 5]
```diff
@@ -70,21 +76,18 @@ const Str = {
...
         return str.replace(/\n/g, '<br />');
     },
 
     /**
      * Decodes the given HTML encoded string.
      *
-     * @param {String} s The string to decode.
-     * @return {String} The decoded string.
+     * @param s The string to decode.
+     * @return string The decoded string.
```
Comment (Skalakid) [inline lib/str.ts:88]:
```suggestion
     * @returns The decoded string.
```

File: lib/str.ts [hunk 6]
```diff
@@ -109,31 +112,31 @@ const Str = {
...
         return first === second || this.htmlDecode(first) === second || this.htmlEncode(first) === second;
     },
 
     /**
      * Creates an ID that can be used as an HTML attribute from @p str.
      *
-     * @param {String} str A string to create an ID from.
-     * @return {String} The ID string made from @p str.
+     * @param str A string to create an ID from.
+     * @return string The ID string made from @p str.
```
Comment (Skalakid) [inline lib/str.ts:137]:
```suggestion
     * @returns The ID string made from @p str.
```

File: lib/str.ts [hunk 7]
```diff
@@ -746,106 +725,89 @@ const Str = {
...
      * Replace first N characters of the string with maskChar
      * eg: maskFirstNCharacters( '1234567890', 6, 'X' ) yields XXXXXX7890
-     * @param {String} str string to mask
-     * @param {Number} n number of characters we want to mask from the string
-     * @param {String} mask string we want replace the first N chars with
-     * @return {String} masked string
+     * @param str string to mask
+     * @param num number of characters we want to mask from the string
+     * @param mask string we want replace the first N chars with
+     * @returns masked string
```
Comment (Skalakid) [inline lib/str.ts:765]:
```suggestion
     * @param str String to mask
     * @param num Number of characters we want to mask from the string
     * @param mask String we want replace the first N chars with
     * @returns Masked string
```

File: lib/str.ts [hunk 8]
```diff
@@ -887,14 +847,14 @@ const Str = {
...
      * Checks if a string could be the masked version of another one.
      *
-     * @param {String} first string to compare
-     * @param {String} second string to compare
-     * @param {String} [mask] defaults to X
-     * @return {Boolean} true when first could be the masked version of second
+     * @param strA string to compare
+     * @param strB string to compare
+     * @param [mask] defaults to X
+     * @returns true when first could be the masked version of second
```
Comment (Skalakid) [inline lib/str.ts:853]:
```suggestion
     * @param strA String to compare
     * @param strB String to compare
     * @param [mask] Defaults to X
     * @returns True when first could be the masked version of second
```

File: lib/str.ts [hunk 9]
```diff
@@ -1010,32 +946,30 @@ const Str = {
...
         return matches;
     },
 
     /**
      * A simple GUID generator taken from https://stackoverflow.com/a/32760401/9114791
      *
-     * @param {String} [prefix] an optional prefix to put in front of the guid
-     * @returns {String}
+     * @param [prefix] an optional prefix to put in front of the guid
+     *
```
Comment (Skalakid) [inline lib/str.ts:970]:
```suggestion
```

**Review by Skalakid — COMMENTED** [review]
LGTM, left some small comments connected to function descriptions, to make them have a consistent format

**Review by Skalakid — APPROVED** [review]

File: lib/str.ts [hunk 10]
```diff
@@ -295,21 +289,21 @@ const Str = {
     /**
      * Returns true if the haystack begins with the needle
      *
-     * @param {String} haystack  The full string to be searched
-     * @param {String} needle    The case-sensitive string to search for
-     * @return {Boolean} Retruns true if the haystack starts with the needle.
+     * @param haystack  The full string to be searched
+     * @param needle    The case-sensitive string to search for
+     * @returns Retruns true if the haystack starts with the needle.
```
Comment (VickyStash) [inline lib/str.ts:294]:
```suggestion
     * @param haystack The full string to be searched
     * @param needle  The case-sensitive string to search for
     * @returns True if the haystack starts with the needle.
```

File: lib/str.ts [hunk 11]
```diff
@@ -259,13 +254,13 @@ const Str = {
...
-     * @param {String} input
-     * @returns {Number} The number of bytes used by string
+     * @returns The number of bytes used by string
      */
-    getByteLength(input) {
+    getByteLength(input: string): number {
         // Force string type
         const stringInput = String(input);
         let byteLength = 0;
+        // eslint-disable-next-line @typescript-eslint/prefer-for-of
```
Comment (VickyStash) [inline lib/str.ts:263]: Is it possible to use `for of` instead of ignoring the rule?

File: lib/str.ts [hunk 12]
```diff
@@ -364,74 +358,73 @@ const Str = {
     /**
      * Checks that the string is a domain name (e.g. example.com)
      *
-     * @param {String} string The string to check for domainnameness.
+     * @param str The string to check for domainnameness.
      *
-     * @returns {Boolean} True iff the string is a domain name
+     * @returns True iff the string is a domain name
```
Comment (VickyStash) [inline lib/str.ts:363]:
```suggestion
     * @returns True if the string is a domain name
```

File: lib/str.ts [hunk 13]
```diff
@@ -364,74 +358,73 @@ const Str = {
...
 
     /**
      * Checks that the string is an email address.
      * NOTE: TLDs are not just 2-4 characters. Keep this in sync with _inputrules.php
      *
-     * @param {String} string The string to check for email validity.
+     * @param str The string to check for email validity.
      *
-     * @returns {Boolean} True iff the string is an email
+     * @returns True iff the string is an email
```
Comment (VickyStash) [inline lib/str.ts:384]:
```suggestion
     * @returns True if the string is an email
```

File: lib/str.ts [hunk 14]
```diff
@@ -364,74 +358,73 @@ const Str = {
...
-    areValidEmails(str) {
+    areValidEmails(str: string): boolean {
         const string = this.removeTrailingComma(str);
         if (string === '') {
             return true;
         }
 
         const emails = string.split(',');
         let result = true;
+        // eslint-disable-next-line @typescript-eslint/prefer-for-of
```
Comment (VickyStash) [inline lib/str.ts:427]: Same question

File: lib/str.ts [hunk 15]
```diff
@@ -488,96 +478,87 @@ const Str = {
...
 
     /**
      * Checks if a string's length is in the specified range
      *
-     * @param {String} string The subject
-     * @param {Number} minimumLength
-     * @param {Number} [maximumLength]
+     * @param str The subject
+     * @param minimumLength
+     * @param [maximumLength]
```
Comment (VickyStash) [inline lib/str.ts:550]: Do we really still need this description?

File: lib/str.ts [hunk 16]
```diff
@@ -1070,36 +1003,23 @@ const Str = {
         return website.toLowerCase() + this.cutBefore(match[1], match[2]);
     },
 
-    /**
-     * Checks if parameter is a string or function
-     * if it is a function then we will call it with
-     * any additional arguments.
```
Comment (VickyStash) [inline lib/str.ts:1076]: Maybe we can leave the comment here as well so it's shown to the dev during usage

**Review by VickyStash — APPROVED** [review]

**Comment by carlosmiceli** [comment]
Hey @blazejkustra I'm seeing files missing, and only package.json changed, does that sound right?

**Review by carlosmiceli — APPROVED** [review]

**Comment by aldo-expensify** [comment]
Hmmm, now that `str.js` was migrated to `str.ts`, Web-Expensify complains if I try to update `expensify-common` to the current version;

```
Running "browserify:web" (browserify) task
>> Error: Can't walk dependency graph: Cannot find module 'expensify-common/lib/str' from '/Users/aldocanepa/Expensidev/Web-Expensify/site/shared.js'
>>     required by /Users/aldocanepa/Expensidev/Web-Expensify/site/shared.js
Warning: Error running grunt-browserify. Use --force to continue.
```

I wonder if the Web-Expensify project supports `ts` in its dependencies or not.

**Comment by blazejkustra** [comment]
@aldo-expensify What command do you use to update `expensify-common`? (I think you need to instal it from npm with `npm install expensify-common`)

**Comment by blazejkustra** [comment]
> Does that give a different result compared to what you proposed?

Yes, this is because with TS there is a build step before the package is published to NPM. It was introduced [here](https://github.com/Expensify/expensify-common/pull/699)

**Comment by aldo-expensify** [comment]
ahh, maybe if I had updated the imports in Web-Expensify to look like this it could have worked:

image


Currently the import looks like this:

image

**Pull request merged**